### DISCUSSION TITLE
Use generic type (defined at class level) when implementing interface defining generic methods

### DISCUSSION THREAD

**Discussion by maurosampietro** [dotnet/csharplang#6353] [Language Ideas]
It would be nice to have a way to instruct the compiler to use the generic types defined at class level in a generic method definition; 
This would lead to interfaces being more useful when dealing with coexisting generic and non generic implementations of a type.

Basically avoiding Compiler Warning (level 3) CS0693 and Compiler Error CS0535
and allowing generic/non generic version of a class to be casted to the same interface type.

The point of this feature is to make it possible to write something like this, which now does not compile:

```c#
IJsonSerializer jsonSer = new JsonSerializer();
jsonSer = new JsonSerializer<MyType>();
```

Example:

```c#
    public interface IJsonSerializer
    {
        T Deserialize<T>( string str );
    }

    public sealed class JsonSerializer<T> : IJsonSerializer
        where T : class, new()
    {
        //case 1: if the method is defined this way you get errors about 'not implementing the interface'
        T Deserialize( string str ){...}
        
        //case 2: if the method is defined like this the compiler warns i'm using an existing name for a new generic param for the method.
        T Deserialize<T>( string str ){...}

       //I'm asking if it would make sense to have a keyword (for example 'this') to specify that i want to use the already defined T.
       //Coder experience should be like case 1: no need to pass in the generic param.
       T Deserialize<this T>(string str){...}
    }

   public sealed class JsonSerializer : IJsonSerializer
   {
        T Deserialize<T>( string str ){..}
   }

```

This way my interface in this case could be truly useful.

**Comment by HaloFour** [comment]
In order for `JsonSerializer<T1>` to implement `IJsonSerializer` it is required to implement the method `Deserialize<T2>`.  I intentionally use two different names there because they are two different and completely unrelated generic type parameters, and even if someone creates an instance of `JsonSerializer<DateTime>` that class _must_ support `Deserialize<string>` and `Deserialize<int>` and `Deserialize<object>`, etc.

**Reply by maurosampietro** [reply]
I know it works that way.

My point here is to create a mechanism that force T2 to be exactly T1, avoiding the need to provide the generic type via the method each time the method is called and allowing something like:

```c#
IJsonSerializer jsonSer = new JsonSerializer<MyType>();
jsonSer = new JsonSerializer();

```

**Reply by HaloFour** [reply]
> I know it works that way.
> 
> My point here is to create a mechanism that force T2 to be exactly T1

That can't be forced because that would break the contract.  That class _must_ have a `Deserialize` method that accepts a generic type parameter.

If you wanted you could add a non-generic helper method that doesn't accept the generic type parameter, like this:

```cs
public sealed class JsonSerializer<T> : IJsonSerializer
        where T1 : class, new()
    {

    public T2 Deserialize<T2>(string str) { ... }

    public T1 Deserialize(string str) => this.Deserialize<T1>(str);
}
```

You can also implement the `Deserialize<T>` method explicitly so that it's not directly callable by consumers except through the interface.  But you can't not implement the method.
